framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -528,56 +589,135 @@ private AnnotationMirror createBooleanAnnotationMirror(List<Boolean> values) {
                 }
             }
         }
+
+        // Refine the values of the operands in the store based on the result of the comparison.
+
+        List<Number> thenLeftVals = new ArrayList<>();
+        List<Number> elseLeftVals = new ArrayList<>();
+
+        List<Number> thenRightVals = new ArrayList<>();
+        List<Number> elseRightVals = new ArrayList<>();
+
+        // Because resultValues contains the (boolean) results of each comparison (i.e. the product of the number
+        // of elements in each list), we have to iterate through parts of the list at a time in order to find all
+        // the results of comparisons involving a particular value in the left list.
+        for (int i = 0; i < lefts.size(); i++) {
+            for (int j = 0; j < rights.size(); j++) {
+                if (resultValues.get(j + i * rights.size())) {
+                    thenLeftVals.add(lefts.get(i));
+                    thenRightVals.add(rights.get(j));
+                } else {
+                    elseLeftVals.add(lefts.get(i));
+                    elseRightVals.add(rights.get(j));
+                }
+            }
+        }
+
+        CFStore thenStore = result.getThenStore();
+        CFStore elseStore = result.getElseStore();
+
+        createAnnotationFromResultsAndAddToStore(thenStore, thenLeftVals, leftNode);
+        createAnnotationFromResultsAndAddToStore(elseStore, elseLeftVals, leftNode);
+
+        createAnnotationFromResultsAndAddToStore(thenStore, thenRightVals, rightNode);
+        createAnnotationFromResultsAndAddToStore(elseStore, elseRightVals, rightNode);
+
         return resultValues;
     }
 
+    private void createAnnotationFromResultsAndAddToStore(
+            CFStore store, List<?> results, Node node) {
+        // If a zero-length list is passed, this returns bottom - which is very problematic.
+        AnnotationMirror anno =
+                atypefactory.createResultingAnnotation(
+                        node.getType(), results.size() == 0 ? null : results);
+        AnnotationMirror currentAnno =
+                atypefactory
+                        .getAnnotatedType(node.getTree())
+                        .getAnnotationInHierarchy(atypefactory.BOTTOMVAL);
+        Receiver rec = FlowExpressions.internalReprOf(analysis.getTypeFactory(), node);
+        store.insertValue(
+                rec, atypefactory.getQualifierHierarchy().greatestLowerBound(anno, currentAnno));
```

Comment:
So do you think that we should change the code here? As is, it fails two of the tests in `Refinement2.java` in the dead code sections because it assigns the (actually correct) type to the variable and then finds it inconsistent with the expressed annotation. I'm actually okay with dead code issuing that kind of warning, so I think we should just change `Refinement2.java` to expect those assignments to fail.